java/src/main/native/src/RmmJni.cpp

```diff
@@ -79,18 +90,54 @@ class tracking_resource_adaptor final : public base_tracking_resource_adaptor {
 
   std::size_t get_total_allocated() override { return total_allocated.load(); }
 
+  void push_thread_memory_tracker() override { memory_tracker_stack.emplace(); }
+
+  long pop_thread_memory_tracker() override {
+    auto top_tracker = memory_tracker_stack.top();
+    auto ret = top_tracker.max_outstanding;
+    memory_tracker_stack.pop();
+    if (memory_tracker_stack.empty()) {
+      alloc_map.clear();
+    } else {
+      // carry the max to the next level
+      memory_tracker_stack.top().max_outstanding += ret;
+    }
+    return ret;
+  }
+
 private:
   Upstream *const resource;
   std::size_t const size_align;
   std::atomic_size_t total_allocated{0};
 
+  void thread_allocated(long addr, std::size_t num_bytes) {
+    if (!memory_tracker_stack.empty()) {
+      alloc_map[addr] = num_bytes;
+      memory_tracker &tracker = memory_tracker_stack.top();
+      tracker.current_outstanding += num_bytes;
+      tracker.max_outstanding = std::max(tracker.current_outstanding, tracker.max_outstanding);
+    }
+  }
+
+  void thread_freed(long addr, std::size_t num_bytes) {
+    if (!memory_tracker_stack.empty()) {
+      auto it = alloc_map.find(addr);
+      if (it != alloc_map.end()) {
+        auto tracker = memory_tracker_stack.top();
+        tracker.current_outstanding -= it->second;
+        alloc_map.erase(it);
+      }
+    }
+  }
+
   void *do_allocate(std::size_t num_bytes, rmm::cuda_stream_view stream) override {
     // adjust size of allocation based on specified size alignment
     num_bytes = (num_bytes + size_align - 1) / size_align * size_align;
 
     auto result = resource->allocate(num_bytes, stream);
     if (result) {
       total_allocated += num_bytes;
+      thread_allocated(reinterpret_cast<long>(result), num_bytes);
```

Comment:
`std::uintptr_t` would be the more appropriate type for casting a pointer to an integer.

Reply:
Casting pointer to `long` is a common practice in JNI since we want to store long value in Java (no `uintptr_t`  type in Java). But if we don't return the pointer then of course we should cast to `uintptr_t` for computing intermediate variables.

Reply:
No pointers are being handled in this PR so I'd rather not got to `uintptr_t` in this case.